### ISSUE TITLE
TypeError: 'Call' object is not subscriptable

### ISSUE THREAD

**Issue by jaraco** [xonsh/xonsh#3875]
After copying the `.xonshrc` from my local macOS workstation to a Raspberry Pi Buster with Python 3.9 installed from source, I get the reported error when starting xonsh.

What can I do to troubleshoot?

## xonfig

<details>

```
$ xonfig
+------------------+-----------------+
| xonsh            | 0.9.23          |
| Python           | 3.9.0           |
| PLY              | 3.11            |
| have readline    | True            |
| prompt toolkit   | 3.0.7           |
| shell type       | prompt_toolkit  |
| pygments         | None            |
| on posix         | True            |
| on linux         | True            |
| distro           | unknown         |
| on darwin        | False           |
| on windows       | False           |
| on cygwin        | False           |
| on msys2         | False           |
| is superuser     | False           |
| default encoding | utf-8           |
| xonsh encoding   | utf-8           |
| encoding errors  | surrogateescape |
| on jupyter       | False           |
| jupyter kernel   | None            |
+------------------+-----------------+
```

</details>

## Expected Behavior
`xonsh` should start up without an error or provide an error that's indicative of the issue.

## Current Behavior

I tried bisecting the error and found that if I remove this code and everything after it, the error goes away:

```
def get_repo_url():
	raw = $(git remote get-url --push origin).rstrip()
	return raw.replace('https://github.com/', '')
```

If I remove only that function, the error changes slightly to:

```
TypeError: 'Name' object is not subscriptable
```

The few lines that precede the `get_repo_url` definition look like this:

```
aliases.update(gclip='pbcopy', pclip='pbpaste')

# workaround for https://bugs.python.org/issue22490
${...}.pop('__PYVENV_LAUNCHER__', None)
```

Removing those lines has no effect on the error.

I suspect I've encountered this before, but I don't recall how I fixed it.

### Traceback (if applicable)

<details>

```
jaraco@Katsushiro ~ $ env XONSH_DEBUG=1 XONSH_SHOW_TRACEBACK=1 .local/bin/xonsh
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/home/jaraco/.local/pipx/venvs/xonsh/lib/python3.9/site-packages/xonsh/environ.py", line 2240, in xonsh_script_run_control
    run_script_with_cache(filename, execer, ctx)
  File "/home/jaraco/.local/pipx/venvs/xonsh/lib/python3.9/site-packages/xonsh/codecache.py", line 160, in run_script_with_cache
    ccode = compile_code(filename, code, execer, glb, loc, mode)
  File "/home/jaraco/.local/pipx/venvs/xonsh/lib/python3.9/site-packages/xonsh/codecache.py", line 119, in compile_code
    ccode = execer.compile(code, glbs=glb, locs=loc, mode=mode, filename=filename)
  File "/home/jaraco/.local/pipx/venvs/xonsh/lib/python3.9/site-packages/xonsh/execer.py", line 135, in compile
    tree = self.parse(input, ctx, mode=mode, filename=filename, transform=transform)
  File "/home/jaraco/.local/pipx/venvs/xonsh/lib/python3.9/site-packages/xonsh/execer.py", line 95, in parse
    tree, input = self._parse_ctx_free(input, mode=mode, filename=filename)
  File "/home/jaraco/.local/pipx/venvs/xonsh/lib/python3.9/site-packages/xonsh/execer.py", line 219, in _parse_ctx_free
    tree = self.parser.parse(
  File "/home/jaraco/.local/pipx/venvs/xonsh/lib/python3.9/site-packages/xonsh/parsers/base.py", line 497, in parse
    tree = self.parser.parse(input=s, lexer=self.lexer, debug=debug_level)
  File "/home/jaraco/.local/pipx/venvs/xonsh/lib/python3.9/site-packages/xonsh/ply/ply/yacc.py", line 335, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/home/jaraco/.local/pipx/venvs/xonsh/lib/python3.9/site-packages/xonsh/ply/ply/yacc.py", line 1122, in parseopt_notrack
    p.callable(pslice)
  File "/home/jaraco/.local/pipx/venvs/xonsh/lib/python3.9/site-packages/xonsh/parsers/v38.py", line 519, in p_return_stmt
    value=p[2][0] if p[2] is not None else None,
TypeError: 'Call' object is not subscriptable
error running xonsh run control file '/home/jaraco/.xonshrc': 'Call' object is not subscriptable
```
</details>

## Steps to Reproduce

I've no idea how to reproduce it, except to install Python 3.9 on Raspbian, install `xonsh[ptk]`, and then copy [this profile file](https://www.dropbox.com/s/y5yut89wgph3qf2/.xonshrc?dl=0) to the user's home. I haven't tried reproducing it, as that would take quite some time.

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by jaraco** [comment]
Creating a new xonsh environment under the pre-built Python 3.7 for the platform, the error does not occur, so it seems the issue is related to a difference between Python 3.7 and Python 3.9 but on this platform only. I'm running xonsh under Python 3.9 on my macOS environment without any problems (using the same .xonshrc).

**Comment by anki-code** [comment]
Hi @jaraco! A bit off topic: if you're connecting to Raspberry Pi Buster over ssh you can install/update/remove repeatable xonsh environment using [xxh](https://github.com/xxh/xxh).

**Comment by jaraco** [comment]
> you can try to install/update/remove repeatable xonsh environment using [xxh](https://github.com/xxh/xxh).

Nice. Never heard of it, but looks like a cool concept. I'll give it a try.

**Comment by scopatz** [comment]
Thanks for reporting this issue @jaraco

**Comment by scopatz** [comment]
FWIW this seems to be an issue in constructing the return statement.  It seems like `testlist_star_expr_opt` is not coming back as a string on that platform.  I have no idea why it would succeed on some versions of 3.9 and fail on a different arch

**Comment by scopatz** [comment]
BTW - here is a test for this: #3883

**Comment by jaraco** [comment]
Cool - I'll try running that test on the indicated platform. If that doesn't exhibit the error, I'll continue to pare down the xonshrc until I can get a more surgical reproducer.

**Comment by jaraco** [comment]
Interestingly, running the test suite produces [29 failures](https://gist.github.com/jaraco/a2fd1cd3b4282644294a1166b64141a5) with seemingly related errors like:

```
_________________________________________________ test_str_true_slice _________________________________________________

    def test_str_true_slice():
>       check_ast('"hello"[True,0:3]', False)
E       AssertionError: Ast nodes fields differ : body (of type <class 'ast.Subscript'>) != body (of type <class 'ast.Call'>)
```

My suspicion is that some of these factors are implicated:

- compile options used to build Python from source
- new PEG parser
- AMD64 platform

**Comment by jaraco** [comment]
The instructions I used to [build Python are here](https://installvirtual.com/how-to-install-python-3-8-on-raspberry-pi-raspbian/) (except I used [this source](https://www.python.org/ftp/python/3.9.0/Python-3.9.0.tgz)).

**Comment by scopatz** [comment]
Thanks @jaraco! Yeah, I think we are not fully ready for 3.9 yet

**Comment by laloch** [comment]
I know about those test errors since Python 3.9 beta releases. They are related to the new Python PEG parser. They do not necessarily mean that the Xonsh parser is broken. If you want to get rid of them and any potential bugs induced by the PEG parser, set `PYTHONOLDPARSER=1` in your environment.
@xonsh/xore, we should probably publish the information about `PYTHONOLDPARSER` at some very visible place and/or maybe implement a warning at launch for users who already upgraded to Python 3.9.

**Comment by jaraco** [comment]
I tried running the tests with `PYTHONOLDPARSER=1` and now I get 30 failures (where before I got 29)--basically the same as before. After activating the venv where the dependencies are installed, In bash, I ran: `$ PYTHONOLDPARSER=1 xonsh run-tests.xsh`. I haven't pulled any changes in the local checkout, so the code should be the same.

Similarly, setting the var appears to have no effect on the originally-reported issue.
